From b0b8d7061663a6b0b95f606440a39bd7ce2af3fe Mon Sep 17 00:00:00 2001
Message-ID: <b0b8d7061663a6b0b95f606440a39bd7ce2af3fe.1776946505.git.bertrand.marquis@arm.com>
In-Reply-To: <55d56a174a0f04520d4a1558630bb0b739074b13.1776946505.git.bertrand.marquis@arm.com>
References: <55d56a174a0f04520d4a1558630bb0b739074b13.1776946505.git.bertrand.marquis@arm.com>
From: Bertrand Marquis <bertrand.marquis@arm.com>
Date: Thu, 16 Apr 2026 19:30:46 +0200
Subject: [RFC PATCH 14/28] virtio: add helpers to query and reset virtqueue
 region caches

Add a public helper to query whether a virtqueue region cache is valid.

This gives later virtio-msg code a stable way to check queue-cache state,
while reset continues to use the existing internal helper path.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>
---
 hw/virtio/virtio.c         | 7 +++++++
 include/hw/virtio/virtio.h | 1 +
 2 files changed, 8 insertions(+)

diff --git a/hw/virtio/virtio.c b/hw/virtio/virtio.c
index d30bf80bee8c..55d395ee5a09 100644
--- a/hw/virtio/virtio.c
+++ b/hw/virtio/virtio.c
@@ -262,6 +262,13 @@ static void virtio_virtqueue_reset_region_cache(struct VirtQueue *vq)
     }
 }
 
+bool virtio_queue_region_cache_valid(VirtIODevice *vdev, int n)
+{
+    VirtQueue *vq = &vdev->vq[n];
+
+    return qatomic_read(&vq->vring.caches) != NULL;
+}
+
 void virtio_init_region_cache(VirtIODevice *vdev, int n)
 {
     VirtQueue *vq = &vdev->vq[n];
diff --git a/include/hw/virtio/virtio.h b/include/hw/virtio/virtio.h
index 3a24d109fb74..ae8477569865 100644
--- a/include/hw/virtio/virtio.h
+++ b/include/hw/virtio/virtio.h
@@ -377,6 +377,7 @@ void virtio_queue_get_rings(VirtIODevice *vdev, int n, hwaddr *desc,
                             hwaddr *avail, hwaddr *used);
 void virtio_queue_update_rings(VirtIODevice *vdev, int n);
 void virtio_init_region_cache(VirtIODevice *vdev, int n);
+bool virtio_queue_region_cache_valid(VirtIODevice *vdev, int n);
 void virtio_queue_set_align(VirtIODevice *vdev, int n, int align);
 void virtio_queue_notify(VirtIODevice *vdev, int n);
 uint16_t virtio_queue_vector(VirtIODevice *vdev, int n);
-- 
2.53.0

